src/Features/LanguageServer/ProtocolUnitTests/DocumentChanges/DocumentChangesTests.cs

```diff
@@ -0,0 +1,367 @@
+{
+    public class DocumentChangesTests : AbstractLanguageServerProtocolTests
+    {
+        [Fact]
+        public async Task DidOpen_DocumentIsTracked()
+        {
+            var source =
+@"class A
+{
+    void M()
+    {
+        {|type:|}
+    }
+}";
+            var (workspace, locationTyped, documentText) = await GetWorkspaceAndLocationAsync(source);
+
+            using (workspace)
+            {
+                var queue = CreateRequestQueue(workspace.CurrentSolution);
+
+                await DidOpen(queue, workspace.CurrentSolution, CreateDidOpenTextDocumentParams(locationTyped, documentText));
+
+                var document = queue.GetTestAccessor().GetTrackedDocuments().FirstOrDefault();
+
+                Assert.NotNull(document);
+                Assert.Equal(documentText, document.GetTextSynchronously(default).ToString());
+            }
+        }
+
+        [Fact]
+        public async Task MultipleDidOpen_Errors()
+        {
+            var source =
+@"class A
+{
+    void M()
+    {
+        {|type:|}
+    }
+}";
+            var (workspace, locationTyped, documentText) = await GetWorkspaceAndLocationAsync(source);
+
+            using (workspace)
+            {
+                var queue = CreateRequestQueue(workspace.CurrentSolution);
+
+                await DidOpen(queue, workspace.CurrentSolution, CreateDidOpenTextDocumentParams(locationTyped, documentText));
+
+                await Assert.ThrowsAsync<InvalidOperationException>(() => DidOpen(queue, workspace.CurrentSolution, CreateDidOpenTextDocumentParams(locationTyped, documentText)));
+            }
+        }
+
+        [Fact]
+        public async Task DidCloseWithoutDidOpen_Errors()
+        {
+            var source =
+@"class A
+{
+    void M()
+    {
+        {|type:|}
+    }
+}";
+            var (workspace, locationTyped, documentText) = await GetWorkspaceAndLocationAsync(source);
+
+            using (workspace)
+            {
+                var queue = CreateRequestQueue(workspace.CurrentSolution);
+
+                await Assert.ThrowsAsync<InvalidOperationException>(() => DidClose(queue, workspace.CurrentSolution, CreateDidCloseTextDocumentParams(locationTyped)));
+            }
+        }
+
+        [Fact]
+        public async Task DidChangeWithoutDidOpen_Errors()
+        {
+            var source =
+@"class A
+{
+    void M()
+    {
+        {|type:|}
+    }
+}";
+            var (workspace, locationTyped, documentText) = await GetWorkspaceAndLocationAsync(source);
+
+            using (workspace)
+            {
+                var queue = CreateRequestQueue(workspace.CurrentSolution);
+
+                await Assert.ThrowsAsync<InvalidOperationException>(() => DidChange(queue, workspace.CurrentSolution, CreateDidChangeTextDocumentParams(locationTyped.Uri, (0, 0, "goo"))));
+            }
+        }
+
+        [Fact]
+        public async Task DidClose_DocumentNotTracked()
```

Comment:
name slightly misleading - thought it was going to be an error as we were trying to close an untracked document.  Maybe DidClose_DocumentNoLongerTracked? (or something better, is untracks a word 😆 )